libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -286,7 +334,12 @@ bytes BinaryTransform::run(Module const& _module)
 
 bytes BinaryTransform::operator()(Literal const& _literal)
 {
-	return toBytes(Opcode::I64Const) + lebEncodeSigned(_literal.value);
+	if (holds_alternative<uint32_t>(_literal.value))
+		return toBytes(Opcode::I32Const) + lebEncodeSigned(static_cast<int32_t>(get<uint32_t>(_literal.value)));
+	else if (holds_alternative<uint64_t>(_literal.value))
+		return toBytes(Opcode::I64Const) + lebEncodeSigned(static_cast<int64_t>(get<uint64_t>(_literal.value)));
+	else
+		yulAssert(false, "Invalid literal type");
```

Comment:
Can you use the following or would it not work because of the implicit convertability of the two types?
```suggestion
	return std::visit(GenericVisitor{
		[&](uint32_t _value) -> bytes { return toBytes(Opcode::I32Const) + lebEncodeSigned(_value); },
		[&](uint64_t _value) -> bytes { return toBytes(Opcode::I64Const) + lebEncodeSigned(_value); },
	}, _literal.value);
```

Reply:
No, unfortunately not:
- With `[&](uint32_t _value) ->` it does compile but gives the wrong result (since the implicit `uint32_t` -> `int64_t` cast is still there).
- With `[&](int32_t _value) ->` it does not:
    ```
    error: no type named ‘type’ in ‘struct std::invoke_result<solidity::util::GenericVisitor<solidity::yul::wasm::BinaryTransform::operator()(const solidity::yul::wasm::Literal&)::<lambda(int32_t)>, solidity::yul::wasm::BinaryTransform::operator()(const solidity::yul::wasm::Literal&)::<lambda(int64_t)> >, const unsigned int&>’
    ```

The best I can do is to add explicit casts back:
``` c++
return std::visit(GenericVisitor{
	[&](uint32_t _value) -> bytes { return toBytes(Opcode::I32Const) + lebEncodeSigned(static_cast<int32_t>(_value)); },
	[&](uint64_t _value) -> bytes { return toBytes(Opcode::I64Const) + lebEncodeSigned(static_cast<int64_t>(_value)); },
}, _literal.value);
```

But I think that the cast is not the problem here, just a symptom. Using `lebEncodeSigned()` with unsigned arguments is just tricky. I got burned by it at first and only the error from `wasm-validate` made me realize that I'm using it wrong. I'm tempted to create a separate overload for `lebEncodeSigned(int32_t)` but I didn't do so becuse the implementation would be exactly the same as for `int64_t`. Only the parameter type would differ. It would also probably be ambiguous with regard to implicit casts (though I could just rename it to `lebEncodeSigned32()` and `lebEncodeSigned64()` to step around the issue).

Maybe defining `lebEncodeSigned(variant<uint32_t, uint64_t>)` would be a better idea?

Reply:
looks good with the static cast to me

Reply:
Done. Changed to visitor with casts.